packages/apputils-extension/src/licensesplugin.ts

```diff
@@ -0,0 +1,241 @@
+      ) + '/';
+
+    /**
+     * Create a MainAreaWidget for a license viewer
+     */
+    function createLicenseWidget(args: Licenses.ICreateArgs) {
+      const licensesModel = new Licenses.Model({
+        ...args,
+        licensesUrl,
+        trans,
+        serverSettings: app.serviceManager.serverSettings
+      });
+      const content = new Licenses({ model: licensesModel });
+      content.id = `${licensesNamespace}-${++counter}`;
+      content.title.label = licensesText;
+      content.title.icon = copyrightIcon;
+      const main = new MainAreaWidget({
+        content,
+        reveal: licensesModel.licensesReady
+      });
+
+      main.toolbar.addItem(
+        'refresh-licenses',
+        new CommandToolbarButton({
+          id: CommandIDs.refreshLicenses,
+          args: { noLabel: 1 },
+          commands
+        })
+      );
+
+      main.toolbar.addItem('spacer', Toolbar.createSpacerItem());
+
+      for (const format of Object.keys(Licenses.REPORT_FORMATS)) {
+        const button = new CommandToolbarButton({
+          id: CommandIDs.licenseReport,
+          args: { format, noLabel: 1 },
+          commands
+        });
+        main.toolbar.addItem(`download-${format}`, button);
+      }
+
+      return main;
+    }
+
+    // register license-related commands
+    commands.addCommand(CommandIDs.licenses, {
+      label: licensesText,
+      execute: (args: any) => {
+        // bail if no license API is available from the server
+        if (!PageConfig.getOption('licensesUrl')) {
+          console.warn('No license API available from the server');
+          return;
+        }
+        const licenseMain = createLicenseWidget(args as Licenses.ICreateArgs);
+        shell.add(licenseMain, 'main', { type: 'Licenses' });
+
+        // add to tracker so it can be restored, and update when choices change
+        void licensesTracker.add(licenseMain);
+        licenseMain.content.model.trackerDataChanged.connect(() => {
+          void licensesTracker.save(licenseMain);
+        });
+        return licenseMain;
+      }
+    });
+
+    if (restorer) {
+      void restorer.restore(licensesTracker, {
+        command: CommandIDs.licenses,
+        name: widget => 'licenses',
+        args: widget => {
+          const { currentBundleName, currentPackageIndex, packageFilter } =
+            widget.content.model;
+
+          const args: Licenses.ICreateArgs = {
+            currentBundleName,
+            currentPackageIndex,
+            packageFilter
+          };
+          return args as ReadonlyJSONObject;
+        }
+      });
+    }
+
+    return licensesTracker;
+  }
+};
+
+/**
+ * A plugin to add a licenses reporting tools.
+ */
+export const licensesPlugin: JupyterFrontEndPlugin<void> = {
+  id: '@jupyterlab/apputils-extension:licenses-plugin',
+  description: 'Adds licenses reporting tools.',
+  requires: [ITranslator],
+  optional: [ILicensesTracker, IMainMenu, ICommandPalette],
```

Comment:
It looks to me `licensesTracker` should be a required one, as this plugin does do anything without it?